cip/CIP2016-06-22-nested-subqueries.adoc

```diff
@@ -0,0 +1,383 @@
+--
+
+toc::[]
+
+
+== Motivation
+
+Subqueries - i.e. queries within queries - are a powerful and expressive feature allowing for:
+
+  * Increased query expressivity
+  * Better query construction and readability
+  * Easier query composition and reuse
+  * Post-processing as a single unit results from multiple queries
+  * Perform a sequence of multiple write commands for each record
+
+== Background
+
+This CIP may be viewed in light of the EXISTS CIP and the forthcoming Pattern Comprehension CIP, both of which propose variants of subqueries.
+
+== Proposal
+
+Nested subqueries are self-contained Cypher queries that are run within the scope of an outer Cypher query.
+
+This proposal suggests the introduction of new nested subquery constructs to Cypher.
+
+* Read-only nested match subqueries of the form `MATCH { [(a)->(b) [WHERE ...]] ... RETURN * }`
+* Read-only nested optional match subqueries of the form `OPTIONAL { [(a)->(b) [WHERE ...]] ... RETURN * }`
+* Read-only nested mandatory match subqueries of the form `MANDATORY { [(a)->(b) [WHERE ...]] ... [RETURN *] }`
+* Read/Write nested subqueries of the form `DO WHEN ... { ... }` (not ending with `RETURN`)
+
+All forms are introduced with a keyword in conjunction with optional subclauses which are then followed by an inner query in curly braces.
+
+Nested subqueries may be correlated - i.e. the inner query may use variables from the outer query - or uncorrelated.
+
+Nested subqueries can be contained within other nested subqueries at an arbitrary (but finite) depth.
+
+
+**1. Read-only nested match subqueries**
+
+We propose the addition of new syntax to the `MATCH` clause for expressing read-only nested match subqueries.
+
+A nested read-only match subquery is denoted using the following syntax: `MATCH { <inner-match-query> }`.
+
+The inner match query is a full read-only Cypher query.
+
+Moreover, any valid read-only Cypher query from which the leading `MATCH` keyword has been omitted may also be used as an inner match query.
+
+This rule only applies if the leading `MATCH` clause is the root clause of the inner query (i.e. is not the first clause inside a nested query or a `UNION`).
+
+
+**2. Read-only nested optional match subqueries**
+
+We propose the addition of a new `OPTIONAL` clause for expressing read-only nested optional match subqueries.
+
+A nested read-only optional match subquery is denoted using the following syntax: `OPTIONAL { <inner-match-query> }`.
+
+
+**3. Read-only nested mandatory match subqueries**
+
+We propose the addition of a new `MANDATORY` clause for expressing read-only nested mandatory match subqueries.
+
+A nested read-only mandatory match subquery is denoted using the following syntax: `MANDATORY { <inner-mandatory-query> }`.
+
+The inner mandatory query is any inner match query.
+
+Moreover, any inner match query from which the trailing final `RETURN` clause has been omitted may also be used as an inner mandatory query.
```

Comment:
We're actually thinking about making `RETURN` mandatory on `MANDATORY` match based on the reason that this will be the 95% use case and having to name in the remaining 5% isn't a big deal.